primitives/npos-elections/solution-type/src/from_assignment_helpers.rs

```diff
@@ -0,0 +1,64 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helpers to generate the push code for `from_assignment` implementations. This can be shared
+//! between both single_page and double_page, thus extracted here.
+//!
+//! All of the code in this helper module assumes some variable names, namely `who` and
+//! `distribution`.
+
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+
+pub(crate) fn from_impl_single_push_code() -> TokenStream2 {
+	quote!(push((
+		voter_index(&who).or_invalid_index()?,
+		target_index(&distribution[0].0).or_invalid_index()?,
+	)))
+}
+
+pub(crate) fn from_impl_double_push_code() -> TokenStream2 {
+	quote!(push((
+		voter_index(&who).or_invalid_index()?,
+		(target_index(&distribution[0].0).or_invalid_index()?, distribution[0].1,),
+		target_index(&distribution[1].0).or_invalid_index()?,
+	)))
+}
+
+pub(crate) fn from_impl_rest_push_code(count: usize) -> TokenStream2 {
```

Comment:
sidenote: maybe we can factorize single, double and rest by using an array of length 0, and 1 for single and double.

Reply:
yeah, @emostov also had this suggestion. Arrays of fixed size should have no encoding overhead and it makes the code easier. I'll look into it.

Reply:
Factored out `double` from everywhere, not sure about single. The `[]` in different places looks a bit ugly.